README.md
# datacoco-batch
Integrated integration to Batchy - a batch control system


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.1.2] - 2020-12-10
### Changes
    - Fix conflicts in dependency - requests>=2.20,<=2.22

## [0.1.1] - 2020-03-18
### Changes
    - Support external package future minor versions


## [0.1.0] - 2020-02-06
### Added
    - Contributing guildelines
    - Readme documentation
    - Fix reaadthedocs config
    - Codacy config
    - Initial datacoco batch class

[0.1.2]: https://github.com/equinoxfitness/datacoco-batch/compare/v0.1.1...v0.1.2
[0.1.1]: https://github.com/equinoxfitness/datacoco-batch/compare/v0.1.0...v0.1.1
[0.1.0]: https://github.com/equinoxfitness/datacoco-batch/releases/tag/v0.1.0


tests/test_batch.py
import unittest
from datacoco_batch.batch import Batch


class TestBatch(unittest.TestCase):

    test_wf = "cc3_test_wf"

    @classmethod
    def setUpClass(cls):
        cls.batchy = Batch(
            wf=cls.test_wf,
            server="hypergate-services.equinoxfitness.com",
            port="80",
        )

    def test_open_batch(self):  # "status": "open"
        resp = self.batchy.open()
        status = resp["global"]["status"]
        self.assertEqual(status, "open")

    def test_get_status(self):
        expected_keys = ["global"]
        resp = self.batchy.get_status()
        self.assertEqual(list(resp.keys()), expected_keys)

    def test_close_batch(self):  # "status": "success"
        resp = self.batchy.close()
        status = resp["global"]["status"]

        self.assertEqual(status, "success")

    def test_fail_batch(self):  # "status": "failure"
        resp = self.batchy.fail()
        status = resp["global"]["status"]

        self.assertEqual(status, "failure")

    def test_get_status_summary(self):
        expected_keys = [
            "batch_id",
            "status",
            "failure_cnt",
            "open_cnt",
            "batch_start",
            "batch_end",
        ]
        resp = self.batchy.get_status_summary()
        self.assertEqual(list(resp.keys()), expected_keys)


datacoco_batch/Batch.py
import requests
import os, sys

class Batch:
    """
    integrated integration to Batchy batch control system
    """
    def __init__(self, wf:str, server:str, port:int):
        self.wf = wf
        try:
            server = server
        except:
            server = '0.0.0.0'
        try:
            port = port
        except:
            port = '8050'
        # different URL if behind load balancer
        if "hypergate" in server and port == '80':
            self.server_url = 'http://' + server + '/batchy'
        else:
            self.server_url = 'http://' + server + ':' + port

    def open(self, fmt=None):
        """
        opens batch and returns batch params, by default batch params will be returned as python objects
        :param wf:
        :param fmt: infa, json (default is python object)
        :return:
        """
        r = None
        if fmt == 'json':
            r = requests.get(self.server_url + '/open_batch/' + self.wf).text()
        elif fmt == 'infa':
            r = requests.get(self.server_url + '/open_batch/infa/' + self.wf).text()
        else:
            r = requests.get(self.server_url + '/open_batch/' + self.wf).json()
        return r

    def close(self):
        """
        close batch
        :param wf:
        :return:
        """
        r = requests.get(self.server_url + '/close_batch/' + self.wf).json()
        return r

    def fail(self):
        """
        fail batch
        :param wf:
        :return:
        """
        r = requests.get(self.server_url + '/fail_batch/' + self.wf).json()
        return r

    def get_status(self):
        """
        get status
        :param wf:
        :return:
        """
        r = requests.get(self.server_url + '/get_status/' + self.wf).json()
        return r

    def get_status_summary(self):
        """
        get a single value for the whole workflow
        :return:
        """
        details = self.get_status()
        r = {'batch_id': 0, 'status': 'success', 'failure_cnt': 0, 'open_cnt': 0, 'batch_start': '', 'batch_end': ''}
        for k,x in details.items():
            if  x.get('status') == 'failure':
                r['failure_cnt'] += 1
            if  x.get('status') == 'open':
                r['open_cnt'] += 1
            r['batch_start'] = x.get('batch_start') if r['batch_start'] is None or x.get('batch_start') <  r['batch_start'] else  r['batch_start']
            r['batch_end'] = x.get('batch_end') if 'batch_end' in r or x.get('batch_end') > r['batch_end'] else r['batch_end']
            r['batch_id'] = x.get('batch_id')
        if r['failure_cnt'] > 0:
            r['status'] = 'failure'
        elif r['open_cnt'] > 0:
            r['status'] = 'open'
        return r


datacoco_batch/batch.py
import requests


class Batch:
    """
    integrated integration to Batchy batch control system
    """

    def __init__(self, wf: str, server="0.0.0.0", port="8050"):  # nosec
        self.wf = wf
        # different URL if behind load balancer
        if "hypergate" in server and port == "80":
            self.server_url = "http://" + server + "/batchy"
        else:
            self.server_url = "http://" + server + ":" + port

    def open(self, fmt=None):
        """
        opens batch and returns batch params, by default batch params will be
        returned as python objects
        :param wf:
        :param fmt: infa, json (default is python object)
        :return:
        """
        r = None
        if fmt == "json":
            r = requests.get(self.server_url + "/open_batch/" + self.wf).text()
        elif fmt == "infa":
            r = requests.get(
                self.server_url + "/open_batch/infa/" + self.wf
            ).text()
        else:
            r = requests.get(self.server_url + "/open_batch/" + self.wf).json()
        return r

    def close(self):
        """
        close batch
        :param wf:
        :return:
        """
        r = requests.get(self.server_url + "/close_batch/" + self.wf).json()
        return r

    def fail(self):
        """
        fail batch
        :param wf:
        :return:
        """
        r = requests.get(self.server_url + "/fail_batch/" + self.wf).json()
        return r

    def get_status(self):
        """
        get status
        :param wf:
        :return:
        """
        r = requests.get(self.server_url + "/get_status/" + self.wf).json()
        return r

    def get_status_summary(self):
        """
        get a single value for the whole workflow
        :return:
        """
        details = self.get_status()
        r = {
            "batch_id": 0,
            "status": "success",
            "failure_cnt": 0,
            "open_cnt": 0,
            "batch_start": "",
            "batch_end": "",
        }

        for k, x in details.items():
            if x.get("status") == "failure":
                r["failure_cnt"] += 1
            if x.get("status") == "open":
                r["open_cnt"] += 1
            r["batch_start"] = (
                x.get("batch_start")
                if r["batch_start"] is None
                or x.get("batch_start") < r["batch_start"]
                else r["batch_start"]
            )
            r["batch_end"] = (
                x.get("batch_end")
                if "batch_end" in r or x.get("batch_end") > r["batch_end"]
                else r["batch_end"]
            )
            r["batch_id"] = x.get("batch_id")
        if r["failure_cnt"] > 0:
            r["status"] = "failure"
        elif r["open_cnt"] > 0:
            r["status"] = "open"
        return r


module/module.py
import requests
import os, sys

class Batch(object):
    """
    integrated integration to Batchy batch control system
    """
    def __init__(self, wf:str, config:object):
        self.wf = wf
        try:
            server = config['server']
        except:
            server = '0.0.0.0'
        try:
            port = config['port']
        except:
            port = '8050'
        # different URL if behind load balancer
        if "hypergate" in server and port == '80':
            self.server_url = 'http://' + server + '/batchy'
        else:
            self.server_url = 'http://' + server + ':' + port

    def open(self, fmt=None):
        """
        opens batch and returns batch params, by default batch params will be returned as python objects
        :param wf:
        :param fmt: infa, json (default is python object)
        :return:
        """
        r = None
        if fmt == 'json':
            r = requests.get(self.server_url + '/open_batch/' + self.wf).text()
        elif fmt == 'infa':
            r = requests.get(self.server_url + '/open_batch/infa/' + self.wf).text()
        else:
            r = requests.get(self.server_url + '/open_batch/' + self.wf).json()
        return r

    def close(self):
        """
        close batch
        :param wf:
        :return:
        """
        r = requests.get(self.server_url + '/close_batch/' + self.wf).json()
        return r

    def fail(self):
        """
        fail batch
        :param wf:
        :return:
        """
        r = requests.get(self.server_url + '/fail_batch/' + self.wf).json()
        return r

    def get_status(self):
        """
        get status
        :param wf:
        :return:
        """
        r = requests.get(self.server_url + '/get_status/' + self.wf).json()
        return r

    def get_status_summary(self):
        """
        get a single value for the whole workflow
        :return:
        """
        details = self.get_status()
        r = {'batch_id': 0, 'status': 'success', 'failure_cnt': 0, 'open_cnt': 0, 'batch_start': '', 'batch_end': ''}
        for k,x in details.items():
            if  x.get('status') == 'failure':
                r['failure_cnt'] += 1
            if  x.get('status') == 'open':
                r['open_cnt'] += 1
            r['batch_start'] = x.get('batch_start') if r['batch_start'] is None or x.get('batch_start') <  r['batch_start'] else  r['batch_start']
            r['batch_end'] = x.get('batch_end') if 'batch_end' in r or x.get('batch_end') > r['batch_end'] else r['batch_end']
            r['batch_id'] = x.get('batch_id')
        if r['failure_cnt'] > 0:
            r['status'] = 'failure'
        elif r['open_cnt'] > 0:
            r['status'] = 'open'
        return r
